synapse/storage/client_ips.py

```diff
@@ -75,6 +89,86 @@ def __init__(self, db_conn, hs):
             "before", "shutdown", self._update_client_ips_batch
         )
 
+    @defer.inlineCallbacks
+    def _remove_user_ip_dupes(self, progress, batch_size):
+
+        last_seen_progress = progress.get("last_seen", 0)
+
+        def get_last_seen(txn):
+            txn.execute(
+                """
+                SELECT last_seen FROM user_ips
+                WHERE last_seen > ?
+                ORDER BY last_seen
+                LIMIT 1
+                OFFSET ?
+                """,
+                (last_seen_progress, batch_size)
+            )
+            results = txn.fetchone()
+            return results
+
+        # Get a last seen that's sufficiently far away enough from the last one
+        last_seen = yield self.runInteraction(
+            "user_ips_dups_get_last_seen", get_last_seen
+        )
+
+        if not last_seen:
+            yield self._end_background_update("user_ips_remove_dupes")
+            defer.returnValue(0)
+
+        last_seen = last_seen[0]
+
+        def remove(txn, last_seen_progress, last_seen):
+
+            txn.execute(
+                """
+                SELECT user_id, access_token, ip,
+                       MAX(device_id), MAX(user_agent), MAX(last_seen)
+                FROM (
+                    SELECT user_id, access_token, ip
+                    FROM user_ips
+                    WHERE ? <= last_seen AND last_seen < ?
+                    ORDER BY last_seen
+                ) c
+                INNER JOIN user_ips USING (user_id, access_token, ip)
+                GROUP BY user_id, access_token, ip;
+                HAVING count(*) > 1""",
+                (last_seen_progress, last_seen)
+            )
+            res = txn.fetchall()
```

Comment:
Some quick words about what the fuck this query is doing might be something that other people might appreciate

Reply:
Something like:

> This works by looking at all entries in the given time span, and then for each `(user_id, access_token, ip)` tuple in that range checking for any duplicates in the rest of the table (via a join). It then only returns entries which have duplicates, and the max `last_seen` across all duplicates, which can the be used to delete all other duplicates.  
> It is efficient due to the existence of  `(user_id, access_token, ip)` and `(last_seen)` indices.